complete.py

```diff
@@ -0,0 +1,771 @@
+                the pdf state of the atom subset
+            subset_indices_map : dict
+                dict of {openmm_pdf_state atom_index : openmm_pdf_state_subset atom index}
+            integrator : openmm.Integrator
+                integrator of dynamics
+            ani_handler : ANI1_force_and_energy
+                handler for ani forces and potential energy
+            context_cache : openmmtools.cache.ContextCache, optional
+                The ContextCache to use for Context creation. If None, the global cache
+                openmmtools.cache.global_context_cache is used (default is None).
+            reassign_velocities : bool, optional
+                If True, the velocities will be reassigned from the Maxwell-Boltzmann
+                distribution at the beginning of the move (default is False).
+            n_restart_attempts : int, optional
+                When greater than 0, if after the integration there are NaNs in energies,
+                the move will restart. When the integrator has a random component, this
+                may help recovering. On the last attempt, the ``Context`` is
+                re-initialized in a slower process, but better than the simulation
+                crashing. An IntegratorMoveError is raised after the given number of
+                attempts if there are still NaNs. 
+            reporter : coddiwomple.openmm.reporter.OpenMMReporter, default None
+                a reporter object to write trajectories
+        """
+        super().__init__(openmm_pdf_state,
+                 integrator,
+                 context_cache,
+                 reassign_velocities,
+                 n_restart_attempts)
+        #create a pdf state for the subset indices (usually a vacuum system)
+        self.pdf_state_subset = openmm_pdf_state_subset
+        assert self.pdf_state_subset.temperature == self.pdf_state.temperature, f"the temperatures of the pdf states do not match"
+        
+        #create a dictionary for subset indices
+        self._subset_indices_map = subset_indices_map
+        
+        #create an ani handler attribute that can be referenced
+        self.ani_handler = ani_handler
+        
+        #create a context for the subset atoms that can be referenced
+        self.context_subset, _ = cache.global_context_cache.get_context(self.pdf_state_subset)
+        
+        #create a reporter for the accumulated works
+        self._state_works = {}
+        self._state_works_counter = 0
+        
+        #create a reporter
+        self._write_trajectory = False if reporter is None else True
+        self.reporter=reporter
+        if self._write_trajectory:
+            from coddiwomple.particles import Particle
+            self.particle = Particle(0)
+        else:
+            self.particle = None
+        
+    def _before_integration(self, *args, **kwargs):
+        particle_state = args[0] #define the particle state
```

Comment:
Is this a safe way to pass stuff in to the function? Is there a reason that it's not `_before_integration(particle_state, n_iterations,**kwargs)` ?

Reply:
yes; i don't want to break the `super` class's API (defined in `coddiwomple.openmm.propagators`); in order to make the `Propagator.apply` method amenable to generalized SMC (not necessarily AIS) i need to allow the most flexibility in the `_before_integration` and `_during_integration` as possible, so i pass general args and kwargs